Improve delegation processing in 5.4.4.2:11.2

awwad opened the pull request:
This addresses items 2 and 4 in #35.

Allow multi-role delegation processing to handle role thresholds
and the terminating setting.

Also adjust language to make single-role delegations clearer
about behavior when the image metadata is found and if it cannot be
found.

mnm678 commented on uptane-standard.md:
```diff
@@ -867,9 +867,17 @@ In order to perform full verification, an ECU SHALL perform the following steps:
+                2. If the minimum number of roles in agreement threshold is not met:
+                    1. If the multi-role delegation was terminating, raise an error indicating that image info could not be verified.
+                    2. If the multi-role delegation was not terminating, return no image info and allow the process to proceed to the next delegation.
+        2. If the role specified by the delegation contains signed metadata about the image, return the metadata to be checked and continue to step 11.3.
```
The wording here is a bit confusing. Is continuing to step 11.3 just the result of the return?

awwad replied:
Yeah, it's the presumptive result of the return.  I was trying to soften "skip to 11.3", but maybe we should just exclude it entirely.  What's clearer?

mnm678 replied:
I think removing it would make it more clear.

awwad commented:
Rebased to resolve conflicts.

awwad commented:
@patrickvacek 
Good points.  I think the last commit addresses your comments.  Does that work better?  (And did I introduce anything cringe-worthy in the process? 😄)

patrickvacek commented on uptane-standard.md:
```diff
@@ -867,11 +867,19 @@ In order to perform full verification, an ECU SHALL perform the following steps:
+                    2. If this multi-role delegation is not terminating, return that no image metadata was found (and continue 11.2 with the next delegation).
+        3. If the role specified by the delegation contains signed metadata about the image, return the metadata.
+        4. If the role specified by the delegation does not contain signed metadata about the image:
+            1. If this is a terminating delegation, abort the search for this image, ending the recursion and returning an error code indicating that verified imagine metadata could not be found.
```
typo: "imagine" --> "image" :)

patrickvacek commented on uptane-standard.md:
```diff
@@ -867,11 +867,19 @@ In order to perform full verification, an ECU SHALL perform the following steps:
+            2. Inspect the return values from each:
+                1. If at least a number of roles equal to the minimum number of roles in agreement for this multi-role delegation return the same non-custom metadata (i.e. length and hashes) for the image, then return that metadata for the image.  If the values returned are such that there are multiple sets of non-custom metadata meeting the threshold (e.g. if the threshold is 2/4 and two sets of ...
+                2. If the threshold for the minimum number of roles in agreement is not met:
+                    1. If this multi-role delegation is terminating, abort the search for this image, ending the recursion and returning an error code indicating that verified image metadata could not be found.
```
For the terminating cases, I suggest replacing the comma after "abort the search for this image" with the word "by".

awwad replied:
Fixed, thanks.

awwad commented:
I think the TAP3 side is sound, but more review is more good; it's a little complex.  @tkfu?  @trishankatdatadog?

trishankatdatadog commented on uptane-standard.md:
```diff
@@ -867,11 +867,19 @@ In order to perform full verification, an ECU SHALL perform the following steps:
+        2. If it is a multi-role delegation {{TAP-3}}, then:
+            1. First, recurse (11.2) for each role listed as if there is a single-role delegation to each.
+            2. Inspect the return values from each:
+                1. If at least a number of roles equal to the minimum number of roles in agreement for this multi-role delegation return the same non-custom metadata (i.e. length and hashes) for the image, then return that metadata for the image.  If the values returned are such that there are multiple sets of non-custom metadata meeting the threshold (e.g. if the minimum number of roles in agree ...
```
The last sentence could be simplified, it's a bit hard to read

Otherwise LGTM abeit a bit complicated

awwad replied:
Does this read better?

*If at least a number of roles equal to the minimum number of roles in agreement for this multi-role delegation return the same non-custom metadata (i.e. length and hashes) for the image, then return that metadata for the image.  A conflict exists if there are multiple different sets of non-custom metadata with enough roles in agreement about them to meet the role threshold (for example, if the minimum number of roles in agreement is two and two sets of at least two roles agree on different metadata).  To resolve such a conflict, choose the metadata returned by the set of roles that includes the earliest role in the multi-role delegation.*

trishankatdatadog replied:
Sounds good, thanks for making TAP 3 much clearer!
